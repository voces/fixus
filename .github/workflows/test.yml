name: test

on:
  push:
    paths:
      - '**.ts'
      - 'package.json'

jobs:
  build:

    runs-on: ubuntu-latest

    steps:
    - uses: actions/checkout@v2
    - name: Runs tests
      run: |
        npm i
        npm test
