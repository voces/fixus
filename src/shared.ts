
// This file should not import anything from this repo except libs

import { addScriptHook, W3TS_HOOK } from "@voces/w3ts";
import { AbilityRangePreload } from "./misc/abilityPreload";

export const getterSetterFunc = <T>( init?: T ): ( newValue?: T ) => T => {

	let value = init;
	return ( newValue?: T ): T => {

		if ( newValue !== undefined )
			value = newValue;

		if ( value === undefined )
			throw new Error( "variable left undefined" );

		return value;

	};

};

type GAME_STATES = "init" | "start" | "play";

export const fillArray = <T>( size: number, value: T, arr: Array<T> = [] ): Array<T> => {

	for ( let i = 0; i < size; i ++ )
		arr.push( value );

	return arr;

};

export const fillArrayFn = <T>( size: number, fn: ( index: number ) => T, arr: Array<T> = [] ): Array<T> => {

	for ( let i = 0; i < size; i ++ )
		arr[ i ] = fn( i );

	return arr;

};

export const color = [
	"|CFFFF0303", "|CFF0042FF", "|CFF1CE6B9", "|CFF540081",
	"|CFFFFFF01", "|CFFFE8A0E", "|CFF20C000", "|CFFE55BB0",
	"|CFF959697", "|CFF7EBFF1", "|CFF106246", "|CFF4E2A04",
	"|CFF3F81F8", "|CFFC00040", "|CFFD9D919",
];
export const gameState: ( newState?: GAME_STATES ) => GAME_STATES = getterSetterFunc( "init" as GAME_STATES );
export const goldFactor: ( newFactory?: number ) => number = getterSetterFunc( 1 );
export const isHere = (): boolean => GetPlayerSlotState( GetFilterPlayer() ) === PLAYER_SLOT_STATE_PLAYING;
export const saveskills: Array<number> = fillArray( bj_MAX_PLAYERS, 0 );
export const wws: Array<unit> = [];
export const sheeps: Array<unit> = [];
export const sheepTeam = CreateForce();
export const wispTeam = CreateForce();
export const wolfTeam = CreateForce();
export const wolves: Array<unit> = [];
export const WISP_TYPE = FourCC( "eC01" );

export const BLACK_WOLF_TYPE = FourCC( "E002" );
export const IMBA_WOLF_TYPE = FourCC( "E000" );
export const WHITE_WOLF_TYPE = FourCC( "eC16" );
export const SHEEP_TYPE = FourCC( "uC04" );
export const WOLF_TYPE = FourCC( "EC03" );
export const CLOAK_TYPE = FourCC( "clfm" );
export const DOLLY_TYPE = FourCC( "nshf" );
export const wisps: Array<unit> = [];
export const isSandbox = getterSetterFunc( false );

let someInteger: number;

const countHereEnum = (): void => {

	if ( GetPlayerSlotState( GetEnumPlayer() ) === PLAYER_SLOT_STATE_PLAYING )
		someInteger = someInteger + 1;

};

// Counts players in force that are here
export const countHere = ( f: force ): number => {

	someInteger = 0;
	ForForce( f, countHereEnum );
	return someInteger;

};

const countHereRealEnum = (): void => {

	if ( GetPlayerSlotState( GetEnumPlayer() ) === PLAYER_SLOT_STATE_PLAYING && GetPlayerController( GetEnumPlayer() ) === MAP_CONTROL_USER )
		someInteger = someInteger + 1;

};

// Counts players in force that are here
export const countHereReal = ( f: force ): number => {

	someInteger = 0;
	ForForce( f, countHereRealEnum );
	return someInteger;

};

export const DisplayTimedText = ( duration: number, message: string ): void => {

	let i = 0;

	while ( true ) {

		if ( i === bj_MAX_PLAYERS ) break;
		DisplayTimedTextToPlayer( Player( i ), 0, 0, duration, message );
		i = i + 1;

	}

};

export const TriggerRegisterPlayerChatEventAll = ( t: trigger, s: string, match: boolean ): void => {

	let i = 0;

	while ( true ) {

		if ( i === bj_MAX_PLAYERS ) break;
		TriggerRegisterPlayerChatEvent( t, Player( i ), s, match );
		i = i + 1;

	}

};

export const TriggerRegisterPlayerEventAll = ( t: trigger, e: playerevent ): void => {

	let i = 0;

	while ( true ) {

		if ( i === bj_MAX_PLAYERS ) break;
		TriggerRegisterPlayerEvent( t, Player( i ), e );
		i = i + 1;

	}

};

export const TriggerRegisterPlayerUnitEventAll = ( t: trigger, p: playerunitevent, b: boolexpr ): void => {

	let i = 0;

	while ( true ) {

		if ( i === bj_MAX_PLAYERS ) break;
		TriggerRegisterPlayerUnitEvent( t, Player( i ), p, b );
		i = i + 1;

	}

};

// Grabs the player's main unit
export const mainUnit = ( p: player ): unit => {

	const playerId = GetPlayerId( p );

	if ( IsPlayerInForce( p, wolfTeam ) )
		return wolves[ playerId ];

	if ( IsPlayerInForce( p, sheepTeam ) )
		return sheeps[ playerId ];

	return wisps[ playerId ];

};

export const wolfUnit = ( p: player ): unit => {

	if ( ! IsPlayerInForce( p, wolfTeam ) ) throw new Error( "used wolfUnit on non-wolf" );

	const playerId = GetPlayerId( p );
	const wolf = wolves[ playerId ];

	if ( GetUnitTypeId( wolf ) === WHITE_WOLF_TYPE )
		return wws[ playerId ];

	return wolf;

};

export const SmallText = ( amount: number, u: unit, cc: number, x: number, y: number ): void => {

	if ( GetUnitAbilityLevel( u, FourCC( "Alv1" ) ) <= 0 && IsVisibleToPlayer( GetUnitX( u ), GetUnitY( u ), GetLocalPlayer() ) ) {

		const tt = CreateTextTag();
		SetTextTagPermanent( tt, false );
		SetTextTagPos( tt, GetUnitX( u ) + x, GetUnitY( u ) + y, 25 );
		SetTextTagText( tt, `${color[ cc ]}+${amount}`, 0.0276 );
		SetTextTagColor( tt, 217, 217, 25, 0 );
		SetTextTagFadepoint( tt, 0 );
		SetTextTagVelocity( tt, 0, 0.027734375 );
		SetTextTagLifespan( tt, 3 );

	}

};

addScriptHook( W3TS_HOOK.MAIN_BEFORE, (): void => {

	SetMapFlag( MAP_SHARED_ADVANCED_CONTROL, true );

	ForceEnumAllies( sheepTeam, Player( 0 ), Condition( isHere ) );
	ForceEnumAllies( wolfTeam, Player( 11 ), Condition( isHere ) );

} );

export const BLACK_SHEEP_TYPE = FourCC( "uC02" );
export const SILVER_SHEEP_TYPE = FourCC( "u000" );
export const GOLD_SHEEP_TYPE = FourCC( "u001" );

const sheepTypes = [ SHEEP_TYPE, BLACK_SHEEP_TYPE, SILVER_SHEEP_TYPE, GOLD_SHEEP_TYPE ];
const wolfTypes = [ WOLF_TYPE, BLACK_WOLF_TYPE, IMBA_WOLF_TYPE ];

export const isUnitSheep = ( unit: unit ): boolean =>
	sheepTypes.includes( GetUnitTypeId( unit ) );

export const isUnitWolf = ( unit: unit ): boolean =>
	wolfTypes.includes( GetUnitTypeId( unit ) );

addScriptHook( W3TS_HOOK.MAIN_AFTER, (): void => {

	AbilityRangePreload( FourCC( "A001" ), FourCC( "A00Q" ) );

} );
