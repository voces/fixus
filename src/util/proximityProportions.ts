
import {
	BLACK_WOLF_TYPE,
	IMBA_WOLF_TYPE,
	mainUnit,
	sheeps,
	sheepTeam,
	SmallText,
	wisps,
	wispTeam,
	WOLF_TYPE,
	wolfTeam,
	wolfUnit,
} from "shared";
import { forEachPlayer, reducePlayerUnits } from "util/temp";
import { adjustPlayerGold } from "resources/goldPerSecond";
import { isPlayingPlayer, isComputer } from "./player";

export const GOLEM_TYPE = FourCC( "ewsp" );
const STALKER_TYPE = FourCC( "nfel" );

type Point = {
    x: number;
    y: number;
}

type Amount = {
    gold?: number;
    lumber?: number;
    experience?: number;
}

const UNIT_FACTORS: Record<number, number> = {
	[ WOLF_TYPE ]: 1,
	[ BLACK_WOLF_TYPE ]: 1,
	[ IMBA_WOLF_TYPE ]: 1,
	[ GOLEM_TYPE ]: 0.75,
	[ STALKER_TYPE ]: 0.75,
};
const unitFactor = ( unit: unit ): number => {

	const baseFactor = UNIT_FACTORS[ GetUnitTypeId( unit ) ];

	// no entry or hardcoded 0
	if ( ! baseFactor || baseFactor <= 0 ) return 0;

	// illusions give 75%
	if ( IsUnitIllusion( unit ) ) return baseFactor * 0.75;

	return baseFactor;

};

/**
 * Returns an array where all values sum to 1.
 */
export const normalize = ( arr: number[] ): number[] => {

	const sum = arr.reduce( ( sum, v ) => sum + v, 0 );
	return arr.map( v => v / sum );

};

/**
 * Returns a sigmoid function using the passed constants.
 * @param v Controls the impact of inflection.
 * @param b Linear transformer on `t`.
 * @param c Controls how far t=0 is along the curve.
 */
const sigmoid = ( v: number, b: number, c: number ):
( ( t: number ) => number ) => {

	const cTetrated = c ** c;
	const k = ( cTetrated + 1 ) ** v;
	return ( t: number ): number =>
		k / ( cTetrated + Math.exp( b * t ) ) ** v;

};

const sheepSigmoid = sigmoid( 0.04, 0.03, 4 );

const sheepProximityProportions = (
	{ x, y }: Point,
	amounts: Amount,
): Map<player, Amount> => {

	// calculate proportions, which is the max of the players' units' proportion
	const players: player[] = [];
	let proportions: number[] = [];
	forEachPlayer( p => {

		if (
			! IsPlayerInForce( p, sheepTeam ) ||
			! isPlayingPlayer( p ) && ! isComputer( p )
		) return;

		const xDelta = GetUnitX( sheeps[ GetPlayerId( p ) ] ) - x;
		const yDelta = GetUnitY( sheeps[ GetPlayerId( p ) ] ) - y;
		const distance = ( xDelta ** 2 + yDelta ** 2 ) ** 0.5;
		const proportion = sheepSigmoid( distance );

		proportions.push( proportion );
		players.push( p );

	} );
	proportions = normalize( proportions );

	// calculate amounts
	const proportionAmounts: Map<player, Amount> = new Map();
	let remainders = { gold: 0, lumber: 0, experience: 0 };
	for ( let i = 0; i < players.length; i ++ ) {

		const reals = {
			gold: ( amounts.gold || 0 ) * proportions[ i ] + remainders.gold,
			lumber: ( amounts.lumber || 0 ) * proportions[ i ] + remainders.lumber,
			experience: ( amounts.experience || 0 ) * proportions[ i ] + remainders.experience,
		};

		const integers = {
			gold: Math.round( reals.gold ),
			lumber: Math.round( reals.lumber ),
			experience: Math.round( reals.experience ),
		};

		remainders = {
			gold: integers.gold - reals.gold,
			lumber: integers.lumber - reals.lumber,
			experience: integers.experience - reals.experience,
		};

		proportionAmounts.set( players[ i ], integers );

	}

	return proportionAmounts;

};

const wispSigmoid = sigmoid( 0.08, 0.08, 8 );

const wispProximityProportions = (
	{ x, y }: Point,
	amounts: Amount,
): Map<player, Amount> => {

	// calculate proportions, which is the max of the players' units' proportion
	const players: player[] = [];
	let proportions: number[] = [];
	forEachPlayer( p => {

		if (
			! IsPlayerInForce( p, wispTeam ) ||
			! isPlayingPlayer( p ) && ! isComputer( p )
		) return;

		const xDelta = GetUnitX( wisps[ GetPlayerId( p ) ] ) - x;
		const yDelta = GetUnitY( wisps[ GetPlayerId( p ) ] ) - y;
		const distance = ( xDelta ** 2 + yDelta ** 2 ) ** 0.5;
		const proportion = wispSigmoid( distance );

		proportions.push( proportion );
		players.push( p );

	} );
	proportions = normalize( proportions );

	// calculate amounts
	const proportionAmounts: Map<player, Amount> = new Map();
	let remainders = { gold: 0, lumber: 0, experience: 0 };
	for ( let i = 0; i < players.length; i ++ ) {

		const reals = {
			gold: ( amounts.gold || 0 ) * proportions[ i ] + remainders.gold,
			lumber: ( amounts.lumber || 0 ) * proportions[ i ] + remainders.lumber,
			experience: ( amounts.experience || 0 ) * proportions[ i ] + remainders.experience,
		};

		const integers = {
			gold: Math.round( reals.gold ),
			lumber: Math.round( reals.lumber ),
			experience: Math.round( reals.experience ),
		};

		remainders = {
			gold: integers.gold - reals.gold,
			lumber: integers.lumber - reals.lumber,
			experience: integers.experience - reals.experience,
		};

		proportionAmounts.set( players[ i ], integers );

	}

	return proportionAmounts;

};

const wolfSheepKillSigmoid = sigmoid( 0.01, 0.02, 4 );
const wolfOtherKillSigmoid = sigmoid( 0.04, 0.04, 4 );

// wolves shouldn't have too many units, so fine to iterate over them all
const wolfProximityProportions = (
	{ x, y }: Point,
	amounts: Amount,
	killType: "sheep" | "other" = "other",
): Map<player, Amount> => {

	// calculate proportions, which is the max of the players' units' proportion
	const players: player[] = [];
	let proportions: number[] = [];
	forEachPlayer( p => {

		if ( ! IsPlayerInForce( p, wolfTeam ) ) return;

		const proportion = reducePlayerUnits( p, ( max, unit ) => {

			const xDelta = GetUnitX( unit ) - x;
			const yDelta = GetUnitY( unit ) - y;
			const distance = ( xDelta ** 2 + yDelta ** 2 ) ** 0.5;
			const proportion = ( killType === "sheep" ?
				wolfSheepKillSigmoid :
				wolfOtherKillSigmoid
			)( distance ) * unitFactor( unit );

			if ( proportion > max )	return proportion;
			return max;

		}, 0 );

		proportions.push( proportion );
		players.push( p );

	} );
	proportions = normalize( proportions );

	// calculate amounts
	const proportionAmounts: Map<player, Amount> = new Map();
	let remainders = { gold: 0, lumber: 0, experience: 0 };
	for ( let i = 0; i < proportions.length; i ++ ) {

		const reals = {
			gold: ( amounts.gold || 0 ) * proportions[ i ] + remainders.gold,
			lumber: ( amounts.lumber || 0 ) * proportions[ i ] + remainders.lumber,
			experience: ( amounts.experience || 0 ) * proportions[ i ] + remainders.experience,
		};

		const integers = {
			gold: Math.round( reals.gold ),
			lumber: Math.round( reals.lumber ),
			experience: Math.round( reals.experience ),
		};

		remainders = {
			gold: reals.gold - integers.gold,
			lumber: reals.lumber - integers.lumber,
			experience: reals.experience - integers.experience,
		};

		proportionAmounts.set( players[ i ], integers );

	}

	return proportionAmounts;

};

export const proximityProportions = (
	origin: Point,
	amounts: Amount,
	killer: player,
	killType: "sheep" | "other" = "other",
): Map<player, Amount> => {

	if ( IsPlayerInForce( killer, wolfTeam ) )
		return wolfProximityProportions( origin, amounts, killType );

	if ( IsPlayerInForce( killer, sheepTeam ) )
		return sheepProximityProportions( origin, amounts );

	return wispProximityProportions( origin, amounts );

};

export const awardBounty = (
	origin: Point,
	amounts: Amount,
	killer: player,
	killType: "sheep" | "other" = "other",
): void => {

	const bounties = proximityProportions( origin, amounts, killer, killType );

	bounties.forEach( ( bounty, player ) => {

		let offsets = 0;

		if ( bounty.gold && bounty.gold > 0 ) {

			adjustPlayerGold( player, bounty.gold );
			SmallText( bounty.gold, mainUnit( player ), "gold", offsets * 16, offsets * - 64 );
			offsets ++;

		}

		if ( bounty.experience && bounty.experience > 0 ) {

			const unit = wolfUnit( player );

			if ( ! unit ) return;

			AddHeroXP( unit, bounty.experience, true );
			SmallText( bounty.experience, unit, "purple", offsets * 16, offsets * - 64 );
			offsets ++;

		}

		if ( bounty.lumber && bounty.lumber > 0 ) {

			const unit = wolfUnit( player );

			if ( ! unit ) return;

			AddHeroXP( unit, bounty.lumber, true );
			SmallText( bounty.lumber, unit, "darkgreen", offsets * 16, offsets * - 64 );
			offsets ++;

		}

	} );

};
