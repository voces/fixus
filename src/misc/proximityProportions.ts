
import {
	BLACK_WOLF_TYPE,
	IMBA_WOLF_TYPE,
	sheeps,
	sheepTeam,
	WOLF_TYPE,
	wolfTeam,
} from "../shared";
import { forEachPlayer, reducePlayerUnits } from "../util/temp";

export const GOLEM_TYPE = FourCC( "ewsp" );
const STALKER_TYPE = FourCC( "nfel" );

type Point = {
    x: number;
    y: number;
}

type Amount = {
    gold?: number;
    lumber?: number;
    experience?: number;
}

const WIDTH = 5376 * 2;

const SHEEP_DISTANCE_FACTOR = 1 / 27;
const SHEEP_DENOM = ( WIDTH ** 2 * 2 - 128 ) ** SHEEP_DISTANCE_FACTOR;

const WOLF_DISTANCE_FACTOR = 1 / 9;
const WOLF_DENOM = ( WIDTH ** 2 * 2 - 128 ) ** WOLF_DISTANCE_FACTOR;

const UNIT_FACTORS: Record<number, number> = {
	[ WOLF_TYPE ]: 1,
	[ BLACK_WOLF_TYPE ]: 1,
	[ IMBA_WOLF_TYPE ]: 1,
	[ GOLEM_TYPE ]: 0.5,
	[ STALKER_TYPE ]: 0.5,
};
const unitFactor = ( unit: unit ): number => {

	const baseFactor = UNIT_FACTORS[ GetUnitTypeId( unit ) ];

	// no entry or hardcoded 0
	if ( ! baseFactor || baseFactor <= 0 ) return 0;

	// illusions give 50%
	if ( IsUnitIllusion( unit ) ) return baseFactor * 0.50;

	return baseFactor;

};

export const normalize = ( arr: number[] ): number[] => {

	const sum = arr.reduce( ( sum, v ) => sum + v, 0 );
	return arr.map( v => v / sum );

};

const sheepProximityProportions = (
	{ x, y }: Point,
	amounts: Amount,
): Map<player, Amount> => {

	// calculate proportions, which is the max of the players' units' proportion
	const players: player[] = [];
	let proportions: number[] = [];
	forEachPlayer( p => {

		if ( ! IsPlayerInForce( p, sheepTeam ) ) return;

		const distanceSquared = ( GetUnitX( sheeps[ GetPlayerId( p ) ] ) - x ) ** 2 + ( GetUnitY( sheeps[ GetPlayerId( p ) ] ) - y ) ** 2;
		const proportion = 1 - Math.max( distanceSquared - 128, 0 ) ** SHEEP_DISTANCE_FACTOR / SHEEP_DENOM;

		proportions.push( proportion );
		players.push( p );

	} );
	proportions = normalize( proportions );

	// calculate amounts
	const proportionAmounts: Map<player, Amount> = new Map();
	let remainders = { gold: 0, lumber: 0, experience: 0 };
	for ( let i = 0; i < players.length; i ++ ) {

		const reals = {
			gold: ( amounts.gold || 0 ) * proportions[ i ] + remainders.gold,
			lumber: ( amounts.lumber || 0 ) * proportions[ i ] + remainders.lumber,
			experience: ( amounts.experience || 0 ) * proportions[ i ] + remainders.experience,
		};

		const integers = {
			gold: Math.round( reals.gold ),
			lumber: Math.round( reals.lumber ),
			experience: Math.round( reals.experience ),
		};

		remainders = {
			gold: integers.gold - reals.gold,
			lumber: integers.lumber - reals.lumber,
			experience: integers.experience - reals.experience,
		};

		proportionAmounts.set( players[ i ], integers );

	}

	return proportionAmounts;

};

// wolves shouldn't have too many units, so fine to iterate over them all
const wolfProximityProportions = (
	{ x, y }: Point,
	amounts: Amount,
): Map<player, Amount> => {

	// calculate proportions, which is the max of the players' units' proportion
	const players: player[] = [];
	let proportions: number[] = [];
	forEachPlayer( p => {

		if ( ! IsPlayerInForce( p, wolfTeam ) ) return;

		const proportion = reducePlayerUnits( p, ( max, unit ) => {

			const distanceSquared = ( GetUnitX( unit ) - x ) ** 2 + ( GetUnitY( unit ) - y ) ** 2;
			const proportion = ( 1 - Math.max( distanceSquared - 128, 0 ) ** WOLF_DISTANCE_FACTOR / WOLF_DENOM ) * unitFactor( unit );
			if ( proportion > max )	return proportion;
			return max;

		}, 0 );

		proportions.push( proportion );
		players.push( p );

	} );
	proportions = normalize( proportions );

	// calculate amounts
	const proportionAmounts: Map<player, Amount> = new Map();
	let remainders = { gold: 0, lumber: 0, experience: 0 };
	for ( let i = 0; i < proportions.length; i ++ ) {

		const reals = {
			gold: ( amounts.gold || 0 ) * proportions[ i ] + remainders.gold,
			lumber: ( amounts.lumber || 0 ) * proportions[ i ] + remainders.lumber,
			experience: ( amounts.experience || 0 ) * proportions[ i ] + remainders.experience,
		};

		const integers = {
			gold: Math.round( reals.gold ),
			lumber: Math.round( reals.lumber ),
			experience: Math.round( reals.experience ),
		};

		remainders = {
			gold: reals.gold - integers.gold,
			lumber: reals.lumber - integers.lumber,
			experience: reals.experience - integers.experience,
		};

		proportionAmounts.set( players[ i ], integers );

	}

	return proportionAmounts;

};

export const proximityProportions = (
	origin: Point,
	amounts: Amount,
	killer: player,
): Map<player, Amount> => {

	if ( IsPlayerInForce( killer, wolfTeam ) )
		return wolfProximityProportions( origin, amounts );

	return sheepProximityProportions( origin, amounts );

};
