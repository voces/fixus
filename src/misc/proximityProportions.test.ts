
import "../test/w3api";
import { normalize, proximityProportions, GOLEM_TYPE } from "./proximityProportions";
import { wolfTeam, WOLF_TYPE, sheepTeam, sheeps, SHEEP_TYPE } from "../shared";

describe( "normalize", () => {

	it( "sample 1", () => expect( normalize(
		[ 1, 1, 1, 1, 1 ],
	) ).toEqual(
		[ 1 / 5, 1 / 5, 1 / 5, 1 / 5, 1 / 5 ],
	) );

	it( "sample 2", () => expect( normalize(
		[ 1, 2, 3, 4, 5 ],
	) ).toEqual(
		[ 1 / 15, 2 / 15, 3 / 15, 4 / 15, 5 / 15 ],
	) );

} );

describe( "proximityProportions", () => {

	describe( "sheep", () => {

		const withTempSheep = ( arr: number[], fn: ( () => void ) ): void => {

			const oldSheep: unit[] = [];
			arr.forEach( pId => {

				oldSheep[ pId ] = sheeps[ pId ];
				ForceAddPlayer( sheepTeam, Player( pId ) );
				sheeps[ pId ] = CreateUnit( Player( pId ), SHEEP_TYPE, 64 * 1.5 ** pId, 64 * 1.5 ** pId, 270 );

			} );

			fn();

			arr.forEach( pId => {

				RemoveUnit( sheeps[ pId ] );
				sheeps[ pId ] = oldSheep[ pId ];

			} );

		};

		const sheepPlayerIds = [ 0, 1, 3, 4, 6, 7, 9, 10 ];

		for ( let i = 0; i < sheepPlayerIds.length; i ++ )
			it( `with ${i + 1} sheep`, () => withTempSheep( sheepPlayerIds.slice( 0, i + 1 ), () => {

				const entries = Array.from( proximityProportions(
					{ x: 0, y: 0 },
					{ gold: 100 },
					Player( 0 ),
				).entries() );

				expect( entries.map( v => [ v[ 0 ].playerId, v[ 1 ] ] ) ).toMatchSnapshot();

			} ) );

	} );

	describe( "wolves", () => {

		let wolf: unit;

		beforeEach( () => {

			ForceAddPlayer( wolfTeam, Player( 8 ) );
			wolf = CreateUnit( Player( 8 ), WOLF_TYPE, 256, 256, 270 );

		} );

		it( "solo", () => {

			const entries = Array.from( proximityProportions(
				{ x: 0, y: 0 },
				{ gold: 125, experience: 100 },
				Player( 8 ),
			).entries() );

			expect( entries.map( v => [ v[ 0 ].playerId, v[ 1 ] ] ) )
				.toEqual( [[ 8, { gold: 125, experience: 100, lumber: 0 } ]] );

		} );

		it( "even", () => {

			ForceAddPlayer( wolfTeam, Player( 9 ) );
			const wolf2 = CreateUnit( Player( 9 ), WOLF_TYPE, - 256, - 256, 270 );

			const entries = Array.from( proximityProportions(
				{ x: 0, y: 0 },
				{ gold: 125, experience: 100 },
				Player( 8 ),
			).entries() );

			ForceRemovePlayer( wolfTeam, Player( 9 ) );
			RemoveUnit( wolf2 );

			expect( entries.map( v => [ v[ 0 ].playerId, v[ 1 ] ] ) )
				.toEqual( [
					[ 8, { gold: 63, experience: 50, lumber: 0 } ],
					[ 9, { gold: 62, experience: 50, lumber: 0 } ],
				] );

		} );

		it( "spread", () => {

			ForceAddPlayer( wolfTeam, Player( 9 ) );
			const wolf2 = CreateUnit( Player( 9 ), WOLF_TYPE, 1024, 1024, 270 );
			ForceAddPlayer( wolfTeam, Player( 10 ) );
			const wolf3 = CreateUnit( Player( 10 ), WOLF_TYPE, 4096, 4096, 270 );
			ForceAddPlayer( wolfTeam, Player( 11 ) );
			const golem = CreateUnit( Player( 11 ), GOLEM_TYPE, 1024, 1024, 270 );

			const entries = Array.from( proximityProportions(
				{ x: 0, y: 0 },
				{ gold: 125, experience: 100 },
				Player( 8 ),
			).entries() );

			ForceRemovePlayer( wolfTeam, Player( 9 ) );
			RemoveUnit( wolf2 );
			ForceRemovePlayer( wolfTeam, Player( 10 ) );
			RemoveUnit( wolf3 );
			ForceRemovePlayer( wolfTeam, Player( 11 ) );
			RemoveUnit( golem );

			expect( entries.map( v => [ v[ 0 ].playerId, v[ 1 ] ] ) )
				.toEqual( [
					[ 8, { gold: 48, experience: 38, lumber: 0 } ], // old : 35.2%
					[ 9, { gold: 35, experience: 28, lumber: 0 } ], // old : 21.6%
					[ 10, { gold: 16, experience: 13, lumber: 0 } ], // old: 21.6%
					[ 11, { gold: 26, experience: 21, lumber: 0 } ], // old: 21.6%
				] );

		} );

		afterEach( () => {

			ForceRemovePlayer( wolfTeam, Player( 8 ) );
			RemoveUnit( wolf );

		} );

	} );

} );
