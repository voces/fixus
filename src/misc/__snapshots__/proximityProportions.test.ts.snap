// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`proximityProportions sheep with 1 sheep 1`] = `
Array [
  Array [
    0,
    Object {
      "experience": 0,
      "gold": 100,
      "lumber": 0,
    },
  ],
]
`;

exports[`proximityProportions sheep with 2 sheep 1`] = `
Array [
  Array [
    0,
    Object {
      "experience": 0,
      "gold": 52,
      "lumber": 0,
    },
  ],
  Array [
    1,
    Object {
      "experience": 0,
      "gold": 49,
      "lumber": 0,
    },
  ],
]
`;

exports[`proximityProportions sheep with 3 sheep 1`] = `
Array [
  Array [
    0,
    Object {
      "experience": 0,
      "gold": 37,
      "lumber": 0,
    },
  ],
  Array [
    1,
    Object {
      "experience": 0,
      "gold": 35,
      "lumber": 0,
    },
  ],
  Array [
    3,
    Object {
      "experience": 0,
      "gold": 30,
      "lumber": 0,
    },
  ],
]
`;

exports[`proximityProportions sheep with 4 sheep 1`] = `
Array [
  Array [
    0,
    Object {
      "experience": 0,
      "gold": 29,
      "lumber": 0,
    },
  ],
  Array [
    1,
    Object {
      "experience": 0,
      "gold": 27,
      "lumber": 0,
    },
  ],
  Array [
    3,
    Object {
      "experience": 0,
      "gold": 23,
      "lumber": 0,
    },
  ],
  Array [
    4,
    Object {
      "experience": 0,
      "gold": 21,
      "lumber": 0,
    },
  ],
]
`;

exports[`proximityProportions sheep with 5 sheep 1`] = `
Array [
  Array [
    0,
    Object {
      "experience": 0,
      "gold": 25,
      "lumber": 0,
    },
  ],
  Array [
    1,
    Object {
      "experience": 0,
      "gold": 23,
      "lumber": 0,
    },
  ],
  Array [
    3,
    Object {
      "experience": 0,
      "gold": 19,
      "lumber": 0,
    },
  ],
  Array [
    4,
    Object {
      "experience": 0,
      "gold": 18,
      "lumber": 0,
    },
  ],
  Array [
    6,
    Object {
      "experience": 0,
      "gold": 15,
      "lumber": 0,
    },
  ],
]
`;

exports[`proximityProportions sheep with 6 sheep 1`] = `
Array [
  Array [
    0,
    Object {
      "experience": 0,
      "gold": 22,
      "lumber": 0,
    },
  ],
  Array [
    1,
    Object {
      "experience": 0,
      "gold": 21,
      "lumber": 0,
    },
  ],
  Array [
    3,
    Object {
      "experience": 0,
      "gold": 18,
      "lumber": 0,
    },
  ],
  Array [
    4,
    Object {
      "experience": 0,
      "gold": 16,
      "lumber": 0,
    },
  ],
  Array [
    6,
    Object {
      "experience": 0,
      "gold": 13,
      "lumber": 0,
    },
  ],
  Array [
    7,
    Object {
      "experience": 0,
      "gold": 11,
      "lumber": 0,
    },
  ],
]
`;

exports[`proximityProportions sheep with 7 sheep 1`] = `
Array [
  Array [
    0,
    Object {
      "experience": 0,
      "gold": 21,
      "lumber": 0,
    },
  ],
  Array [
    1,
    Object {
      "experience": 0,
      "gold": 20,
      "lumber": 0,
    },
  ],
  Array [
    3,
    Object {
      "experience": 0,
      "gold": 17,
      "lumber": 0,
    },
  ],
  Array [
    4,
    Object {
      "experience": 0,
      "gold": 15,
      "lumber": 0,
    },
  ],
  Array [
    6,
    Object {
      "experience": 0,
      "gold": 12,
      "lumber": 0,
    },
  ],
  Array [
    7,
    Object {
      "experience": 0,
      "gold": 10,
      "lumber": 0,
    },
  ],
  Array [
    9,
    Object {
      "experience": 0,
      "gold": 6,
      "lumber": 0,
    },
  ],
]
`;

exports[`proximityProportions sheep with 8 sheep 1`] = `
Array [
  Array [
    0,
    Object {
      "experience": 0,
      "gold": 20,
      "lumber": 0,
    },
  ],
  Array [
    1,
    Object {
      "experience": 0,
      "gold": 19,
      "lumber": 0,
    },
  ],
  Array [
    3,
    Object {
      "experience": 0,
      "gold": 16,
      "lumber": 0,
    },
  ],
  Array [
    4,
    Object {
      "experience": 0,
      "gold": 14,
      "lumber": 0,
    },
  ],
  Array [
    6,
    Object {
      "experience": 0,
      "gold": 11,
      "lumber": 0,
    },
  ],
  Array [
    7,
    Object {
      "experience": 0,
      "gold": 10,
      "lumber": 0,
    },
  ],
  Array [
    9,
    Object {
      "experience": 0,
      "gold": 7,
      "lumber": 0,
    },
  ],
  Array [
    10,
    Object {
      "experience": 0,
      "gold": 5,
      "lumber": 0,
    },
  ],
]
`;
